api/app/Http/Controllers/AuthController.php

```diff
@@ -115,6 +129,81 @@ public function token(Request $request, UserRepository $userRepository)
 
         $token = $this->jwtAuth->fromUser($user);
 
-        return $this->responder->setTransformer(App::make(AuthTokenTransformer::class))->item($token);
+        return $this->responder->setTransformer($this->app->make(AuthTokenTransformer::class))->item($token);
+    }
+
+    /**
+     * Redirect the user to the Provider authentication page.
+     *
+     * @param  string     $provider
+     * @param  Socialite  $socialite
+     * @return Response
+     */
+    public function redirectToProvider($provider, Socialite $socialite)
+    {
+        $this->validateProvider($provider);
+
+        return $socialite->with($provider)->stateless()->redirect();
+    }
+
+    /**
+     * Obtain the user information from Provider.
+     *
+     * @param  string          $provider
+     * @param  Socialite       $socialite
+     * @param  UserRepository  $repository
+     * @return Response
+     */
+    public function handleProviderCallback($provider, Socialite $socialite, UserRepository $repository)
+    {
+        $this->validateProvider($provider);
+
+        $socialUser = $socialite->with($provider)->stateless()->user();
+
+        // Verify so we received an email address, if using oAuth credentials
+        // with Twitter for instance, that isn't whitelisted, no email
+        // address will be returned with the response.
+        // See the notes in Spira API doc under Social Login for more info.
+        if (!$socialUser->email) {
+            // The app is connected with the service, but the 3rd party service
+            // is not configured or allowed to return email addresses, so we
+            // can't process the data further. Let's throw an exception.
+            throw new UnprocessableEntityException('User object has no email');
```

Comment:
This scenario should be logged as a critical error as it means users can't log in. See http://lumen.laravel.com/docs/errors#logging